tfhe/src/integer/server_key/radix_parallel/tests_signed.rs

```diff
@@ -1474,6 +1518,342 @@ fn integer_signed_default_absolute_value(param: impl Into<PBSParameters>) {
     }
 }
 
+fn integer_signed_default_left_shift<P>(param: P)
+where
+    P: Into<PBSParameters>,
+{
+    let (cks, mut sks) = KEY_CACHE.get_from_params(param);
+    let cks = RadixClientKey::from((cks, NB_CTXT));
+
+    sks.set_deterministic_pbs_execution(true);
+
+    let mut rng = rand::thread_rng();
+
+    let modulus = (cks.parameters().message_modulus().0.pow(NB_CTXT as u32) / 2) as i64;
+
+    assert!(modulus > 0);
+    assert!((modulus as u64).is_power_of_two());
+    let nb_bits = modulus.ilog2() + 1; // We are using signed numbers
+
+    for _ in 0..NB_TEST_SMALLER {
+        let mut clear = rng.gen::<i64>() % modulus;
+
+        let offset = random_non_zero_value(&mut rng, modulus);
+
+        let mut ct = cks.encrypt_signed(clear);
+        sks.unchecked_scalar_add_assign(&mut ct, offset);
+        clear = signed_add_under_modulus(clear, offset, modulus);
+
+        // case when 0 <= shift < nb_bits
+        {
+            // Here we create a encrypted shift value in range O..nb_bits
+            // in a way that the shift ciphertext is seen as having non empty carries
+            let clear_shift = rng.gen::<u32>() % nb_bits;
+            let tmp = cks.encrypt(clear_shift);
+            let mut shift = cks.encrypt(0u32);
+            sks.unchecked_add_assign(&mut shift, &tmp);
+            assert!(!shift.block_carries_are_empty());
+
+            let ct_res = sks.left_shift_parallelized(&ct, &shift);
+            let dec_res: i64 = cks.decrypt_signed(&ct_res);
+            let clear_res = signed_left_shift_under_modulus(clear, clear_shift, modulus);
+            assert_eq!(
+                clear_res, dec_res,
+                "Invalid left shift result, for '{clear} << {clear_shift}', \
+                expected:  {clear_res}, got: {dec_res}"
+            );
+
+            let ct_res2 = sks.left_shift_parallelized(&ct, &shift);
+            assert_eq!(ct_res, ct_res2, "Failed determinism check");
+        }
+
+        // case when shift >= nb_bits
+        {
+            // Here we create a encrypted shift value >= nb_bits
+            // in a way that the shift ciphertext is seen as having non empty carries
+            let mut clear_shift = rng.gen::<u32>() % nb_bits;
+            let tmp = cks.encrypt(clear_shift);
+            let mut shift = cks.encrypt(nb_bits);
+            sks.unchecked_add_assign(&mut shift, &tmp);
+            clear_shift += nb_bits;
+            assert!(!shift.block_carries_are_empty());
+
+            let ct_res = sks.left_shift_parallelized(&ct, &shift);
+            let dec_res: i64 = cks.decrypt_signed(&ct_res);
+            // When nb_bits is not a power of two
+            // then the behaviour is not the same
+            let mut nb_bits = modulus.ilog2();
```

Comment:
can you elaborate a bit ?

Reply:
It comes from how barrel shifter is/are made (so thats also the case for unsigned radix)

If you want to shift an integer type with N bits, the barrel shift only consider the log2(N) bits from the value that gives the shift amount.

This works well when N is a power of two as it is the case for 8, 16, 32, 64, 128.
e.g for a u8 -> 8 bits -> log2(8) = 4, barrel shifter only consider bits in range 0..4 (so 3 bits), with 3 bits we can only give input shift in range 0..=7 -> great u8 has 8 bits
u16 -> 16 bits -> bits in range 0..4 considered -> possible shift values 0..15.
Basically the barrel shift naturally does `shift_amount % nb_bits` if N is a power of two.

for u12 (4 blocks of 3_3)  -> log2(12) = 3.58 -> we have to ceil() -> 4 -> bits in range 0..4 -> shift values 0..15 which is > 12.  Which means that the behaviour is not the same if the input shift value is not already in the good range